src/pudl_archiver/archivers/ferc/xbrl.py

```diff
@@ -299,6 +302,66 @@ async def archive_taxonomies(
     )
 
 
+async def _download_filing(
+    filing: FeedEntry,
+    session: aiohttp.ClientSession,
+) -> bytes:
+    """Download a single filing."""
+    # Download filing
+    response = await retry_async(
+        session.get,
+        args=[str(filing.download_url)],
+        kwargs={"raise_for_status": True},
+    )
+    return await retry_async(response.content.read)
+
+
+async def _download_filings(
+    archive: zipfile.ZipFile,
+    year: Year,
+    filings: set[FeedEntry],
+    form: FercForm,
+    session: aiohttp.ClientSession,
+):
+    """Download all filings for a single year/form.
+
+    Args:
+        archive: ZipFile to write filings to.
+        year: Year to archive.
+        filings: Set of filings indexed from RSS feed.
+        form: Ferc form.
+        session: Async http client session.
+    """
+    metadata = defaultdict(list)
+    for filing in tqdm(
+        sorted(filings, key=lambda f: f.download_url),
+        desc=f"FERC {form.value} {year} XBRL",
+    ):
+        # Download filing
+        response_bytes = await _download_filing(filing, session)
+
+        # Write to zipfile
+        filename = f"{filing.title}_form{filing.ferc_formname.as_int()}_{filing.ferc_period}_{round(filing.published_parsed.timestamp())}.xbrl".replace(
+            " ", "_"
+        )
+        filing_name = f"{filing.title}{filing.ferc_period}"
+        filing_metadata = FilingMetadata.from_rss_metadata(
+            filing, filename, response_bytes
+        )
+        metadata[filing_name].append(filing_metadata.model_dump())
+
+        add_to_archive_stable_hash(
+            archive=archive, filename=filename, data=response_bytes
+        )
+    return {
+        filename: sorted(
+            metadata[filename],
+            key=lambda filing: filing["rss_metadata"]["download_url"],
+        )
+        for filename in sorted(metadata)
+    }
+
+
 async def archive_year(
```

Comment:
I've split this into a couple sub-functions to make it easier to test and I think improve readability.